contracts/Strategy.sol

```diff
@@ -17,95 +17,389 @@ import {
+     * the Strategy is. The Strategy's debt limit is configured in the Vault.
+     *
+     * NOTE: `_debtPayment` should be less than or equal to `_debtOutstanding`.
+     *       It is okay for it to be less than `_debtOutstanding`, as that
+     *       should only used as a guide for how much is left to pay back.
+     *       Payments should be made to minimize loss from slippage, debt,
+     *       withdrawal fees, etc.
+     *
+     * See `vault.debtOutstanding()`.
+     */
     function prepareReturn(uint256 _debtOutstanding)
         internal
         override
         returns (
             uint256 _profit,
             uint256 _loss,
             uint256 _debtPayment
-        )
-    {
-        // TODO: Do stuff here to free up any returns back into `want`
-        // NOTE: Return `_profit` which is value generated by all positions, priced in `want`
-        // NOTE: Should try to free up at least `_debtOutstanding` of underlying position
+    ){
+        _profit = 0;
+        _loss = 0; //for clarity. also reduces bytesize
+        _debtPayment = 0; //for clarity. also reduces bytesize
+		
+        // Pay debt if any
+        if (_debtOutstanding > 0) {
+            (uint256 _amountFreed, uint256 _reportLoss) = liquidatePosition(_debtOutstanding);
+            _debtPayment = _amountFreed > _debtOutstanding? _debtOutstanding : _amountFreed;
+            _loss = _reportLoss;
+        }
+		
+        // Claim profit
+        MMFarmingPool(mmFarmingPool).withdraw(mmFarmingPoolId, 0);
+        uint256 _want = want.balanceOf(address(this));
+        _disposeOfMM();
+        _profit = want.balanceOf(address(this)).sub(_want);
+				
+        return (_profit, _loss, _debtPayment);
     }
 
-    function adjustPosition(uint256 _debtOutstanding) internal override {
-        // TODO: Do something to invest excess `want` tokens (from the Vault) into your positions
-        // NOTE: Try to adjust positions so that `_debtOutstanding` can be freed up on *next* harvest (not immediately)
+    /**
+     * Perform any adjustments to the core position(s) of this Strategy given
+     * what change the Vault made in the "investable capital" available to the
+     * Strategy. Note that all "free capital" in the Strategy after the report
+     * was made is available for reinvestment. Also note that this number
+     * could be 0, and you should handle that scenario accordingly.
+     *
+     * See comments regarding `_debtOutstanding` on `prepareReturn()`.
+     */
+    function adjustPosition(uint256 _debtOutstanding) internal override{
+	
+        //emergency exit is dealt with in prepareReturn
+        if (emergencyExit) {
+            return;
+        }
+		
+        uint256 _before = IERC20(mmVault).balanceOf(address(this));
+        uint256 _after = _before;
+		
+        uint256 _want = want.balanceOf(address(this));
+        if (_want > _debtOutstanding) {
+            _want = _want.sub(_debtOutstanding);
+			
+            want.safeApprove(mmVault, 0);
```

Comment:
you can do all approves in constructor one time to save gas

Reply:
> 
> 
> Looks fine to me. Needs some testing. In particular need to make sure withdrawal fee is being dealt with correctly.
> 
> We also need to check the mm contracts.

thanks for review. mm vaults/strategy are https://github.com/mushroomsforest/deployment/commit/57e983dfa2a3b788d7573e13ec56dd76ef645304